### ISSUE TITLE
net-snmp v5.8:  trapsink to localhost causes snmpd to fail

### ISSUE THREAD

**Issue by mspncp** [net-snmp/net-snmp#34]
@bvanassche I encountered the same issue as described in

* [sourceforge #2978 - trapsink to localhost causes snmpd to fail](https://sourceforge.net/p/net-snmp/bugs/2978)

on my Gentoo system with with the [net-analyzer/net-snmp-5.8-r1](https://packages.gentoo.org/packages/net-analyzer/net-snmp) package.

It seems to me that the sourceforge ticket was closed by you erroneously without confirming the bug, because [you were not able to reproduce the issue](https://sourceforge.net/p/net-snmp/bugs/2978/#2046) with commit 90837dcec21c.

However, I suspect that it is indeed a bug in v5.8 which has disappeared in the mean time on master, possibly during the refactoring of `netsnmp_sockaddr_in3` in 8169acab64012fe0026e8a47a054e16a4c20621e. (Unfortunately, I was not able to verify my conjecture because I failed  to configure and build a vanilla net-snmp from the master branch.)

Please reconsider the issue using release v5.8 (463235edcc53e2477efe35da717fab292741aab5) and let me know whether it is indeed a bug or a misconfiguration. Since v5.8 is the current release it would be helpful to have an official confirmation of the bug (with possible workarounds), to prevent others from spending the same time to figure out the problem.

## Details

The problem occurred on my Gentoo system with the portage package [net-analyzer/net-snmp-5.8-r1](https://packages.gentoo.org/packages/net-analyzer/net-snmp). Looking at the [gentoo patch set](https://gitweb.gentoo.org/repo/gentoo.git/tree/net-analyzer/net-snmp/files), there seem to be no significant changes to the source code of v5.8 (463235edcc53e2477efe35da717fab292741aab5):

```
/usr/portage/net-analyzer/net-snmp/files # for patch in net-snmp-5.8-*.patch; do echo $patch; diffstat $patch; done
net-snmp-5.8-do-not-conflate-LDFLAGS-and-LIBS.patch
 net-snmp-config.in |    8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)
net-snmp-5.8-my_bool.patch
 snmptrapd_sql.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)
net-snmp-5.8-pcap.patch
 config_os_libs2            |   18 ++++++++++--------
 config_project_with_enable |    5 +++++
 2 files changed, 15 insertions(+), 8 deletions(-)
net-snmp-5.8-tinfo.patch
 config_os_libs2 |    1 +
 1 file changed, 1 insertion(+)
```

### The error case

The attached [/etc/snmp/snmpd.conf](https://github.com/net-snmp/net-snmp/files/3819210/snmpd.conf.txt) contains an `agentAddress` and a `trapsink` declaration:

```
agentAddress  udp:127.0.0.1:161
# ...
trapsink     localhost public
```

According to the documentation, I would expect that the default trapsink port is 162. It turns out however, that port 161 is bound twice and the second call fails with EADDRINUSE:

**strace**
```
~ # strace /usr/sbin/snmpd -f  -f -Dnetsnmp_sockaddr_in -Dnetsnmp_udpbase |& grep 'bind.*INET'
bind(12, {sa_family=AF_INET, sin_port=htons(161), sin_addr=inet_addr("127.0.0.1")}, 16) = 0
bind(11, {sa_family=AF_INET, sin_port=htons(161), sin_addr=inet_addr("127.0.0.1")}, 16) = -1 EADDRINUSE (Address already in use)
```

**net-snmpd.log**
```
# cat /var/log/net-snmpd.log 
netsnmp_sockaddr_in: addr 0x7fffac6cca70, inpeername "localhost", default_target ":162"
netsnmp_sockaddr_in: addr 0x7fffac6cca70, inpeername ":162", default_target "[NIL]"
netsnmp_sockaddr_in: check user service 162
netsnmp_sockaddr_in: return { AF_INET, 0.0.0.0:162 }
netsnmp_sockaddr_in: check user service localhost
netsnmp_sockaddr_in: servname not numeric, check if it really is a destination)
netsnmp_sockaddr_in: check destination localhost
netsnmp_sockaddr_in: hostname (resolved okay)
netsnmp_sockaddr_in: return { AF_INET, 127.0.0.1:162 }
netsnmp_sockaddr_in: addr 0x7fffac6cca80, inpeername "localhost", default_target "[NIL]"
netsnmp_sockaddr_in: check user service localhost
netsnmp_sockaddr_in: servname not numeric, check if it really is a destination)
netsnmp_sockaddr_in: check destination localhost
netsnmp_sockaddr_in: hostname (resolved okay)
netsnmp_sockaddr_in: return { AF_INET, 127.0.0.1:161 }
netsnmp_udpbase: open remote UDP: [127.0.0.1]:162->[0.0.0.0]:0
netsnmp_udpbase: binding socket: 12 to UDP: [0.0.0.0]:0->[127.0.0.1]:161
netsnmp_udpbase: socket 12 bound to UDP: [127.0.0.1]:162->[127.0.0.1]:161
Turning on AgentX master support.
netsnmp_sockaddr_in: addr 0x7fffac6ceed0, inpeername "127.0.0.1:161", default_target ":161"
netsnmp_sockaddr_in: addr 0x7fffac6ceed0, inpeername ":161", default_target "[NIL]"
netsnmp_sockaddr_in: check user service 161
netsnmp_sockaddr_in: return { AF_INET, 0.0.0.0:161 }
netsnmp_sockaddr_in: check user service 161
netsnmp_sockaddr_in: check destination 127.0.0.1
netsnmp_sockaddr_in: hostname (resolved okay)
netsnmp_sockaddr_in: return { AF_INET, 127.0.0.1:161 }
netsnmp_udpbase: open local UDP: [127.0.0.1]:161->[0.0.0.0]:0
netsnmp_udpbase: set IP_PKTINFO
netsnmp_udpbase: binding socket: 11 to UDP: [0.0.0.0]:0->[127.0.0.1]:161
netsnmp_udpbase: failed to bind for clientaddr: 98 Address already in use
Error opening specified endpoint "udp:127.0.0.1:161"
Server Exiting with code 1
```

### Workaround

After adding the port number 162 explicitly to the `trapsink` declaration, the daemon starts up without error:

```
agentAddress  udp:127.0.0.1:161
# ...
trapsink     localhost:162 public
```

**strace**
```
~ # strace /usr/sbin/snmpd -f  -f -Dnetsnmp_sockaddr_in -Dnetsnmp_udpbase |& grep 'bind.*INET'
bind(11, {sa_family=AF_INET, sin_port=htons(161), sin_addr=inet_addr("127.0.0.1")}, 16) = 0
read(15, "grep\0--colour=auto\0bind.*INET\0", 1024) = 30
^C
```

**net-snmpd.log**
```
~ # cat /var/log/net-snmpd.log 
netsnmp_sockaddr_in: addr 0x7ffe33b0d430, inpeername "localhost:162", default_target ":162"
netsnmp_sockaddr_in: addr 0x7ffe33b0d430, inpeername ":162", default_target "[NIL]"
netsnmp_sockaddr_in: check user service 162
netsnmp_sockaddr_in: return { AF_INET, 0.0.0.0:162 }
netsnmp_sockaddr_in: check user service 162
netsnmp_sockaddr_in: check destination localhost
netsnmp_sockaddr_in: hostname (resolved okay)
netsnmp_sockaddr_in: return { AF_INET, 127.0.0.1:162 }
netsnmp_udpbase: open remote UDP: [127.0.0.1]:162->[0.0.0.0]:0
Turning on AgentX master support.
netsnmp_sockaddr_in: addr 0x7ffe33b0f890, inpeername "127.0.0.1:161", default_target ":161"
netsnmp_sockaddr_in: addr 0x7ffe33b0f890, inpeername ":161", default_target "[NIL]"
netsnmp_sockaddr_in: check user service 161
netsnmp_sockaddr_in: return { AF_INET, 0.0.0.0:161 }
netsnmp_sockaddr_in: check user service 161
netsnmp_sockaddr_in: check destination 127.0.0.1
netsnmp_sockaddr_in: hostname (resolved okay)
netsnmp_sockaddr_in: return { AF_INET, 127.0.0.1:161 }
netsnmp_udpbase: open local UDP: [127.0.0.1]:161->[0.0.0.0]:0
netsnmp_udpbase: set IP_PKTINFO
netsnmp_udpbase: binding socket: 11 to UDP: [0.0.0.0]:0->[127.0.0.1]:161
NET-SNMP version 5.8
Received TERM or STOP signal...  shutting down...
```

**Comment by mspncp** [comment]
**correction:** the snmpd.conf which I uploaded already contains the 162 port added:

```
trapsink     localhost:162 public
```

You need to remove it to reproduce the issue.

**Comment by bvanassche** [comment]
We plan to release a version v5.8.1 soon but we do not plan to release a variant of v5.8 with only the aforementioned bug fixed. If you want to use the source code from the Net-SNMP project, please download the latest code from the V5-8-patches branch. If you want to keep using a Gentoo portage package, please report this bug to the Gentoo maintainers.

**Comment by mspncp** [comment]
Thank you @bvanassche for confirming the bug.  I can live with the workaround described above until v5.8.1 will be available on Gentoo. The main reason why I reopened the ticket was because from the conversation on sourceforge, in particular [this post](https://sourceforge.net/p/net-snmp/bugs/2978/#e3de) by the reporter and [your reply](https://sourceforge.net/p/net-snmp/bugs/2978/#2046), I got the impression that the issue was closed without noticing the true nature of the bug and confirming it, because you were not able to reproduce it (anymore). Having an official statement that it's indeed a bug and that it will be fixed in v5.8.1 (and there is a workaround) will help others seeking a solution in the meantime.
